modules/infra/src/main/scala/scaladex/infra/SonatypeClient.scala

```diff
@@ -0,0 +1,112 @@
+    extends SonatypeService
+    with CommonAkkaHttpClient
+    with LazyLogging {
+  private implicit val ec: ExecutionContextExecutor = system.dispatcher
+  private val sonatypeUri = "https://repo1.maven.org/maven2"
+  override val poolClientFlow: Flow[
+    (HttpRequest, Promise[HttpResponse]),
+    (Try[HttpResponse], Promise[HttpResponse]),
+    Http.HostConnectionPool
+  ] =
+    Http()
+      .cachedHostConnectionPoolHttps[Promise[HttpResponse]](
+        "repo1.maven.org",
+        settings = ConnectionPoolSettings("max-open-requests = 32")
+      )
+
+  def getAllArtifactIds(groupId: Artifact.GroupId): Future[Seq[Artifact.ArtifactId]] = {
+    val uri = s"$sonatypeUri/${groupId.mavenUrl}/"
+    val request =
+      HttpRequest(uri = uri)
+
+    for {
+      responseFuture <- queueRequest(request)
+      page <- responseFuture.entity.dataBytes.runFold(ByteString(""))(_ ++ _).map(_.utf8String)
+      artifactIds = JsoupUtils.listDirectories(uri, page)
+      artifactIdsParsed = artifactIds.flatMap(Artifact.ArtifactId.parse)
+    } yield artifactIdsParsed.toSeq
+
+  }
+
+  def getAllVersions(groupId: Artifact.GroupId, artifactId: Artifact.ArtifactId): Future[Seq[SemanticVersion]] = {
+    val uri = s"$sonatypeUri/${groupId.mavenUrl}/${artifactId.value}/"
+    val request = HttpRequest(uri = uri)
+
+    (for {
+      responseFuture <- queueRequest(request)
+      page <- responseFuture.entity.dataBytes.runFold(ByteString(""))(_ ++ _).map(_.utf8String)
+      versions = JsoupUtils.listDirectories(uri, page)
+      versionParsed = versions.flatMap(SemanticVersion.parse)
+    } yield versionParsed.toSeq)
+      .recoverWith {
+        case NonFatal(exception) =>
+          logger.warn(s"failed to retrieve versions from $uri because ${exception.getMessage}")
+          Future.successful(Nil)
+      }
+  }
+
+  override def getPomFile(mavenReference: Artifact.MavenReference): Future[Option[(String, Instant)]] = {
+    val groupIdUrl: String = mavenReference.groupId.replace('.', '/')
+    for {
+      artifactId <- Artifact.ArtifactId.parse(mavenReference.artifactId).toFuture
+      pomUrl = getPomUrl(artifactId, mavenReference.version)
+      uri = s"$sonatypeUri/${groupIdUrl}/${mavenReference.artifactId}/${mavenReference.version}/$pomUrl"
+      request = HttpRequest(uri = uri)
+      responseFuture <- queueRequest(request)
+      res <- getPomFileWithLastModifiedTime(responseFuture)
+    } yield res
+  }
+
+  private def getPomFileWithLastModifiedTime(response: HttpResponse): Future[Option[(String, Instant)]] =
+    response match {
+      case _ @HttpResponse(StatusCodes.OK, headers: Seq[model.HttpHeader], entity, _) =>
+        for {
+          page <- entity.dataBytes.runFold(ByteString(""))(_ ++ _).map(_.utf8String)
+          lastModified <- headers.find(_.is("last-modified")).toFuture.map(_.value())
+          lastModifiedParsed <- Try(format(lastModified)).toFuture
+        } yield Some((page, lastModifiedParsed))
```

Comment:
I find it weird to mix `Future`, `Option` and `Try`. All that happen inside a future will be catched if it fails.

```suggestion
        for (page <- entity.dataBytes.runFold(ByteString(""))(_ ++ _).map(_.utf8String))
        yield {
           headers.find(_.is("last-modified"))
             .map(header => format(header.value)
             .map(page -> _)
        }
```

Reply:
Aslo to get the html file from the entity you can use an akka-http `Unmarshaller`:

```
page  <- Unmarshaller.stringUnmarshaller(entity)
```

Reply:
ok